braindecode/models/spdnet.py

```diff
@@ -0,0 +1,227 @@
+    tril : bool
+        If True, only the lower triangular part of the matrix is used
+
+    References
+    ----------
+    .. [1] Zhiwu Huang and Luc Van Gool, 2016,
+           A Riemannian Network for SPD Matrix Learning
+           AAAI
+    """
+
+    def __init__(self, dim, tril=True, ):
+        super(LogEig, self).__init__()
+        self.tril = tril
+        if self.tril:
+            idx_lower = torch.tril_indices(dim, dim, offset=-1)
+            idx_diag = torch.arange(start=0, end=dim, dtype=torch.long)
+            self.idx = torch.cat((idx_diag[None, :].tile((2, 1)), idx_lower), dim=1)
+        self.dim = dim
+
+    def forward(self, X):
+        return self.embed(logm.apply(X))
+
+    def embed(self, X):
+        if self.tril:
+            x_vec = X[:, self.idx[0], self.idx[1]]
+            x_vec[:, self.dim:] *= 2 ** 0.5
+        else:
+            x_vec = X.flatten(start_dim=1)
+        return x_vec
+
+
+class SPDNet(EEGModuleMixin, nn.Module):
+    """SPDNet from [1]_.
+
+    This class is a SPDNet implementation from [1]_.
+
+    Parameters
+    ----------
+    input_type : str
+        If "raw", the input is a batch of raw EEG signals and
+        a covariance matrix is computed
+        If "cov", the input is a batch of covariance matrices and
+        the covariance matrix is used as input
+    n_chans : int
+        Number of channels
+    subspacedim : int
+        Subspace dimension
+    threshold : float
+        Threshold for the rectified linear unit
+    n_outputs : int
+        Output shape
+    tril : bool
+        If True, only the lower triangular part of the matrix is used
+
+    References
+    ----------
+    .. [1] Zhiwu Huang and Luc Van Gool, 2016,
+           A Riemannian Network for SPD Matrix Learning
+           AAAI
+    """
+    def __init__(
+        self,
+        input_type,
+        n_chans,
+        subspacedim,
+        threshold=1e-4,
+        n_outputs=1,
+        chs_info=None,
+        n_times=None,
+        input_window_seconds=None,
+        sfreq=None,
+        add_log_softmax=False,
+        tril=True,
+    ):
+        super().__init__(
+            n_outputs=n_outputs,
+            n_chans=n_chans,
+            chs_info=chs_info,
+            n_times=n_times,
+            input_window_seconds=input_window_seconds,
+            sfreq=sfreq,
+            add_log_softmax=add_log_softmax,
+        )
+        if input_type == "raw":
+            self.cov = CovLayer()
+        elif input_type == "cov":
+            self.cov = nn.Identity()
+        self.bimap = BiMap(n_chans, subspacedim)
+        self.reeig = ReEig(threshold)
+        self.logeig = LogEig(subspacedim, tril=tril)
+        self.len_last_layer = subspacedim * (subspacedim + 1) // 2 if tril else subspacedim**2
+        self.classifier = torch.nn.Linear(self.len_last_layer, n_outputs)
+
+        if add_log_softmax:
+            self.logsoftmax = nn.LogSoftmax(dim=1)
+        else:
+            self.logsoftmax = nn.Identity()
```

Comment:
```suggestion
```
The goal in the long term is to get rid of the log softmax layer in every model (cf #511). The parameter `add_log_softmax` was introduced to throw a depreciation for the models that had one. But new models should not have a log softmax layer at all.

Reply:
Thanks, I wasn't sure if I needed to put it or not.